File: src/validation.js
import { ERROR_CODES, adapterError } from "./errors.js";

export function assertPlainObject(value, fieldName) {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw adapterError(ERROR_CODES.VALIDATION_ERROR, `${fieldName} must be an object`, {
      details: { field: fieldName }
    });
  }
}

export function assertNonEmptyString(value, fieldName) {
  if (typeof value !== "string" || value.trim().length === 0) {
    throw adapterError(ERROR_CODES.VALIDATION_ERROR, `${fieldName} must be a non-empty string`, {
      details: { field: fieldName }
    });
  }
}

export function assertInteger(value, fieldName) {
  if (!Number.isInteger(value)) {
    throw adapterError(ERROR_CODES.VALIDATION_ERROR, `${fieldName} must be an integer`, {
      details: { field: fieldName }
    });
  }
}

export function assertIdentity(input) {
  assertPlainObject(input, "input");
  assertNonEmptyString(input.tenantId, "input.tenantId");
  assertNonEmptyString(input.userId, "input.userId");
}

File: ai_assist_google_docs_adapter/validation.py
from __future__ import annotations

from typing import Any

from .errors import VALIDATION_ERROR, adapter_error


def assert_plain_object(value: Any, field_name: str) -> None:
    if not isinstance(value, dict):
        raise adapter_error(
            VALIDATION_ERROR,
            f"{field_name} must be an object",
            details={"field": field_name},
        )


def assert_non_empty_string(value: Any, field_name: str) -> None:
    if not isinstance(value, str) or len(value.strip()) == 0:
        raise adapter_error(
            VALIDATION_ERROR,
            f"{field_name} must be a non-empty string",
            details={"field": field_name},
        )


def assert_integer(value: Any, field_name: str) -> None:
    if not isinstance(value, int) or isinstance(value, bool):
        raise adapter_error(
            VALIDATION_ERROR,
            f"{field_name} must be an integer",
            details={"field": field_name},
        )


def assert_identity(input_: dict[str, Any]) -> None:
    assert_plain_object(input_, "input")
    assert_non_empty_string(input_.get("tenantId"), "input.tenantId")
    assert_non_empty_string(input_.get("userId"), "input.userId")
